### PR TITLE
Add Heartbeat Timeout Capabilities in Client Agent

### PR DESCRIPTION
**Pull request by AlJaMa**

### CODE REVIEW

File: src/clientagent/AstronClient.cpp [hunk 1]
```diff
@@ -8,6 +8,7 @@
 #include "config/constraints.h"
 #include "dclass/dc/Class.h"
 #include "dclass/dc/Field.h"
+#include "Util/Timeout.h"
```
Comment (CFSworks) [inline src/clientagent/AstronClient.cpp]: Is case important here? The directory is "util" with a lowercase u, but it's spelled here with an uppercase "U"

File: src/clientagent/AstronClient.cpp [hunk 2]
```diff
@@ -61,13 +70,33 @@ class AstronClient : public Client, public NetworkClient
...
+    ~AstronClient()
+    {
+      if (m_heartbeat_timer != nullptr) {
+        delete m_heartbeat_timer;
+      }
+    }
+
+    void heartbeat_timeout()
+    {  
+    lock_guard<recursive_mutex> lock(m_client_lock);
```
Comment (CFSworks) [inline src/clientagent/AstronClient.cpp]: Indent

File: src/clientagent/AstronClient.cpp [hunk 3]
```diff
@@ -424,6 +461,15 @@ class AstronClient : public Client, public NetworkClient
         }
     }
 
+    // handle_client_heartbeat should ensure this client does not get reset for the current interval.
+    // Handler for CLIENT_HEARTBEAT message
+    virtual void handle_client_heartbeat()
+    {
+      if (m_heartbeat_timer != nullptr) {
```
Comment (CFSworks) [inline src/clientagent/AstronClient.cpp]: Indentation inconsistent with rest of file

File: test/test_clientagent.py [hunk 4]
```diff
@@ -3028,5 +3038,29 @@ def test_get_network_address(self):
...
+
+        # Send a heartbeat
+        dg = Datagram()
+        dg.add_uint16(CLIENT_HEARTBEAT)
+        client.send(dg)
+
+        # The client shouldn't hear anything back.
+        self.expectNone(client)
+
+        time.sleep(0.75)
```
Comment (CFSworks) [inline test/test_clientagent.py]: How does this test pass? heartbeat_timeout is 1000. We should *not* be disconnected after only 0.75s (= 750 ms) of waiting.

File: src/clientagent/AstronClient.cpp [hunk 5]
```diff
@@ -61,13 +70,33 @@ class AstronClient : public Client, public NetworkClient
...
-        m_send_version(send_version_to_client.get_rval(config))
+        m_send_version(send_version_to_client.get_rval(config)),
+        m_heartbeat_timeout(heartbeat_timeout_config.get_rval(config))
     {
         initialize();
     }
 
+    ~AstronClient()
+    {
+      if (m_heartbeat_timer != nullptr) {
```
Comment (CFSworks) [inline src/clientagent/AstronClient.cpp]: Inconsistent indent

File: src/clientagent/AstronClient.cpp [hunk 6]
```diff
@@ -61,13 +70,33 @@ class AstronClient : public Client, public NetworkClient
...
+    lock_guard<recursive_mutex> lock(m_client_lock);
+    send_disconnect(CLIENT_DISCONNECT_NO_HEARTBEAT,
+                    "Server timed out while waiting for heartbeat."); 
+    }
+
     void initialize()
     {
+        //If heartbeat, start the heartbeat timer now.
+        if (m_heartbeat_timeout != 0) {
+            m_heartbeat_timer = new Timeout(m_heartbeat_timeout, std::bind(&AstronClient::heartbeat_timeout,this));
```
Comment (CFSworks) [inline src/clientagent/AstronClient.cpp]: I'm getting *really* picky here, but if you're making edits, put a space before "this"

**Comment by CFSworks** [comment]
Aside from stylistic changes, the only thing I'm wondering about here is the test case's sleep delay. It should wait 750 ms, verify it's still connected, wait 500 ms more, then verify we get a disconnect. Everything else is :+1:

File: test/test_clientagent.py (hunk 4 above)
Comment (AlJaMa) [inline test/test_clientagent.py]: Ah, I see how it passes! The expect is waiting. Do we have any check implemented to check for open connections?

File: src/clientagent/AstronClient.cpp (hunk 6 above)
Comment (bobbybee) [inline src/clientagent/AstronClient.cpp]: Emphasis on _really_

File: src/clientagent/AstronClient.cpp [hunk 7]
```diff
@@ -375,10 +407,12 @@ class AstronClient : public Client, public NetworkClient
             NetworkClient::send_disconnect();
         }
         break;
-        case CLIENT_OBJECT_SET_FIELD: {
```
Comment (bobbybee) [inline src/clientagent/AstronClient.cpp]: This change may or may not be useful, but I don't see why it belongs in this pull request. Am I missing something?

Comment (AlJaMa) [inline src/clientagent/AstronClient.cpp]: Looks like I let a quick formatting change into the PR. No, it is not related to the added functionality.

**Comment by bobbybee** [comment]
At the moment I'm not at a position to run the tests, but :+1: on the code. If this works and passes the tests properly, I'm good with it :)

File: src/clientagent/AstronClient.cpp (hunk 7 above)
Comment (bobbybee) [inline src/clientagent/AstronClient.cpp]: I haven't consulted the style guide, although honestly I think braces looks better (and is probably more correct). Furthermore, be consistent; if you undo this change, also add braces to the heartbeat clause :smile:

File: test/test_clientagent.py (hunk 4 above)
Comment (CFSworks) [inline test/test_clientagent.py]: The best way to check for this would be to have the server send a message down to the client and have the client expect that message. If the connection got killed prematurely, the expect will fail. This also results in no delay added to the test.

**Comment by CFSworks** [comment]
```
======================================================================
FAIL: test_errors (__main__.TestClientAgent)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "../test/test_clientagent.py", line 498, in test_errors
    self.assertDisconnect(client, CLIENT_DISCONNECT_OVERSIZED_DATAGRAM)
  File "../test/test_clientagent.py", line 97, in assertDisconnect
    self.assertTrue(dg is not None, "No datagram received while expecting ClientEject")
AssertionError: No datagram received while expecting ClientEject

----------------------------------------------------------------------
```

**Comment by bobbybee** [comment]
@CFSworks Is heartbeat functionality implemented for all tests? Is it disabled in the CA configuration for the other tests?

**Comment by AlJaMa** [comment]
Heartbeat is disabled by default

**Comment by CFSworks** [comment]
All tests pass for me. :+1:
